content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/commands-example.md

```diff
@@ -0,0 +1,243 @@
+  
+    private interface ExampleContractCommand extends Command { }
+    public static class Create implements ExampleContractCommand { }
+    public static class Update implements ExampleContractCommand { }
+    public static class Delete implements ExampleContractCommand { }
+    
+    @Override
+    
+    public final void verify(UtxoLedgerTransaction transaction) {
+    
+        List<? extends ExampleContractCommand> commands = transaction
+                .getCommands(ExampleContractCommand.class);
+                
+        for (ExampleContractCommand command : commands) {
+            if (command instanceof Create) verifyCreate(transaction);
+            else if (command instanceof Update) verifyUpdate(transaction);
+            else if (command instanceof Delete) verifyDelete(transaction);
+            else throw new IllegalStateException("Unrecognised command type.");
+        }
+    }
+    
+    protected void onVerifyCreate(UtxoLedgerTransaction transaction) { }
+    protected void onVerifyUpdate(UtxoLedgerTransaction transaction) { }
+    protected void onVerifyDelete(UtxoLedgerTransaction transaction) { }    
+    private void verifyCreate(UtxoLedgerTransaction transaction) {
+    
+        // Verify base Create constraints
+        // Then verify additional Create constraints implemented by derived contracts
+        onVerifyCreate(transaction);
+        
+    }
+    private void verifyUpdate(UtxoLedgerTransaction transaction) {
+        // Verify base Update constraints
+        // Then verify additional Update constraints implemented by derived contracts
+        onVerifyUpdate(transaction);
+    }
+    
+    private void verifyDelete(UtxoLedgerTransaction transaction) {
+        // Verify base Delete constraints
+        // Then verify additional Delete constraints implemented by derived contracts
+        onVerifyDelete(transaction);
+    }
+}
+```
+
+Refactoring a contract as shown in the above example allows CorDapp implementors to derive from the contract, allowing additional constraints which will be verified in additional to the constraints specified by the base contract.
+
+{{< note >}}
+There are still some outstanding issues with this design, where this design approach no longer fits the design goals of the system being implemented.
+{{</ note >}}
+
+The problem really lies in the `verify` function; for example:
+
+```kotlin
+public final void verify(UtxoLedgerTransaction transaction) {
+  
+    List<? extends ExampleContractCommand> commands = transaction
+            .getCommands(ExampleContractCommand.class);
+            
+    for (ExampleContractCommand command : commands) {
+        if (command instanceof Create) verifyCreate(transaction);
+        else if (command instanceof Update) verifyUpdate(transaction);
+        else if (command instanceof Delete) verifyDelete(transaction);
+        else throw new IllegalStateException("Unrecognised command type.");
+    }
+}
+```
```

Comment:
```suggestion
```kotlin
public final void verify(UtxoLedgerTransaction transaction) {
    List<? extends ExampleContractCommand> commands = transaction
            .getCommands(ExampleContractCommand.class);
            
    for (ExampleContractCommand command : commands) {
        if (command instanceof Create) verifyCreate(transaction);
        else if (command instanceof Update) verifyUpdate(transaction);
        else if (command instanceof Delete) verifyDelete(transaction);
        else throw new IllegalStateException("Unrecognised command type.");
    }
}
```
```